engine/src/test/java/io/zeebe/engine/processor/StreamProcessorHealthTest.java

```diff
@@ -0,0 +1,271 @@
+
+  @After
+  public void after() {
+    // To exit the processing error loop
+    shouldProcessingThrowException.set(false);
+    shouldFlushThrowException.set(false);
+    shouldFailErrorHandlingInTransaction.set(false);
+  }
+
+  @Test
+  public void shouldBeHealthyOnStart() {
+    // when
+    streamProcessor =
+        streamProcessorRule.startTypedStreamProcessor(
+            (processors, context) ->
+                processors.onEvent(
+                    ValueType.WORKFLOW_INSTANCE,
+                    ELEMENT_ACTIVATING,
+                    mock(TypedRecordProcessor.class)));
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+  }
+
+  @Test
+  public void shouldMarkUnhealthyWhenReprocessingRetryLoop() {
+    // given
+    final long firstPosition =
+        streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
+    streamProcessorRule.writeWorkflowInstanceEventWithSource(ELEMENT_ACTIVATED, 1, firstPosition);
+
+    streamProcessor = getErrorProneStreamProcessor();
+
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+    waitUntil(() -> invocation.get() > 1);
+
+    // when
+    streamProcessorRule.getClock().addTime(HEALTH_CHECK_TICK_DURATION.multipliedBy(1));
+    // give some time for scheduled timers to get executed
+    final int retried = invocation.get();
+    waitUntil(() -> retried < invocation.get());
+    streamProcessorRule.getClock().addTime(HEALTH_CHECK_TICK_DURATION.multipliedBy(1));
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+  }
+
+  @Test
+  public void shouldMarkUnhealthyWhenOnErrorHandlingWriteEventFails() {
+    // given
+    streamProcessor = getErrorProneStreamProcessor();
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+
+    // when
+    shouldFlushThrowException.set(true);
+    streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+  }
+
+  @Test
+  public void shouldMarkUnhealthyWhenProcessingOkWriteEventFails() {
+    // given
+    streamProcessor = getErrorProneStreamProcessor();
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+
+    // when
+    shouldProcessingThrowException.set(false);
+    shouldFlushThrowException.set(true);
+    streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+  }
+
+  @Test
+  public void shouldMarkUnhealthyWhenExceptionErrorHandlingInTransaction() {
+    // given
+    streamProcessor = getErrorProneStreamProcessor();
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+
+    // when
+    shouldFailErrorHandlingInTransaction.set(true);
```

Comment:
i think this is not different to the other processing exception or?

Reply:
When processing fails, it goes to error handling. It should fail again in `errorHandlingInTransaction`. This test checks that particular error path. The other one test when 'errorHandlingInTransaction' succeeds, but the next step `writeEvent` fails.

Reply:
aaah get it - since error events also appended ok makes sense